lib/reek/configuration/app_configuration.rb

```diff
@@ -5,89 +5,115 @@ module Reek
-            raise_error(error.to_s, path)
-          end
+      attr_writer :exclude_paths, :default_directive, :directory_directives
 
-          raise_error('Not a hash', path) unless @configuration.is_a? Hash
-        end
+      # @param source_via [String] - the source of the code inspected
+      # Might be a string, STDIN or Filename / Pathname. We're only interested in the source
+      # when it's coming from file since a directory_directive doesn't make sense
+      # for anything else.
+      # @return [Hash | nil] the configuration for the source or nil
+      def directory_directive_for_source(source_via)
+        return unless source_via
+        base_path = Pathname.new(source_via).split[0]
+        hit = directory_directives.keys.detect { |pathname| pathname.to_s =~ /#{base_path}/ }
+        directory_directives[hit] unless hit.nil?
+      end
 
-        def reset
-          @configuration.clear
-          @exclude_paths = nil
-        end
+      def load(options)
+        configuration_file = ConfigurationFileFinder.find_and_load(options: options)
 
-        def exclude_paths
-          @exclude_paths ||= begin
-            @configuration.fetch(EXCLUDE_PATHS_KEY, []).map do |string|
-              Pathname.new(string.chomp('/'))
+        configuration_file.each do |key, value|
+          case key
+          when EXCLUDE_PATHS_KEY
+            handle_exclude_paths(value)
+          else
+            klass = try_to_recognize_smell_type key
+            if klass
+              handle_default_directive klass, value
+            else
+              handle_directory_directive key, value
             end
           end
         end
+      end
 
-        private
-
-        def for_smell_types
-          @configuration.reject { |key, _value| NON_SMELL_TYPE_KEYS.include?(key) }
+      def handle_exclude_paths(paths)
+        self.exclude_paths = paths.map do |path|
+          pathname = Pathname.new path.chomp('/')
+          raise ArgumentError, "Excluded directory #{path} does not exists" unless pathname.exist?
+          pathname
         end
+      end
 
-        def load_smell_type(name)
-          Reek::Smells.const_get(name)
-        rescue NameError
-          report_problem("\"#{name}\" is not a code smell")
-          nil
-        end
+      def handle_default_directive(klass, config)
+        default_directive[klass] = config
+      end
 
-        def report_problem(reason, path)
-          $stderr.puts "Warning: #{message(reason, path)}"
+      def handle_directory_directive(path, config)
+        pathname = Pathname.new path.chomp('/')
+        unless pathname.exist?
+          raise ArgumentError, "Configuration error: No smell or path for #{pathname}"
         end
-
-        def raise_error(reason, path)
-          raise ConfigFileException, message(reason, path)
+        directory_directives[pathname] = config.each_with_object({}) do |(key, value), hash|
+          hash[Reek::Smells.const_get(key)] = value
         end
```

Comment:
>> #each_with_object always trips me up a bit;

I disagree - I believe, each_with_object is almost always a cleaner solution compared to reduce/inject.
Matz seems to agree, otherwise they probably wouldn't have introduced each_with_object to begin with :smile: 
Since that is highly subjective I'd rather leave it like that.